# JavaScript 异步编程

原文链接：[通用异步编程概念 - 学习 Web 开发 | MDN (mozilla.org)](https://developer.mozilla.org/zh-CN/docs/Learn/JavaScript/Asynchronous/Concepts)

​			 	  [异步JavaScript简介 - 学习 Web 开发 | MDN (mozilla.org)](https://developer.mozilla.org/zh-CN/docs/Learn/JavaScript/Asynchronous/Introducing)

​			 	  [合作异步JavaScript: 超时和间隔 - 学习 Web 开发 | MDN (mozilla.org)](https://developer.mozilla.org/zh-CN/docs/Learn/JavaScript/Asynchronous/Timeouts_and_intervals)

​		 		  [优雅的异步处理 - 学习 Web 开发 | MDN (mozilla.org)](https://developer.mozilla.org/zh-CN/docs/Learn/JavaScript/Asynchronous/Promises)

​				   [async和await:让异步编程更简单 - 学习 Web 开发 | MDN (mozilla.org)](https://developer.mozilla.org/zh-CN/docs/Learn/JavaScript/Asynchronous/Async_await)

## 异步编程概念

通常来说，程序都是顺序执行，同一时刻只会发生一件事。如果一个函数依赖于另一个函数的结果，它只能等待那个函数结束才能继续执行，从用户的角度来说，整个程序才算运行完毕。这是令人沮丧的体验，没有充分利用计算机的计算能力 — 尤其是在计算机普遍都有多核 CPU 的时代，坐在那里等待毫无意义，你完全可以在另一个处理器内核上干其他的工作，同时计算机完成耗时任务的时候通知你。这样你可以同时完成其他工作，这就是**异步编程**的出发点。

## 线程

一个**线程**是一个基本的处理过程，程序用它来完成任务。每个线程一次只能执行一个任务：

```
Task A --> Task B --> Task C
```

JavaScript 是单线程的，JavaScript 传统上是单线程的。即使有多个内核，也只能在单一线程上运行多个任务，此线程称为主线程（**main thread**）。

经过一段时间，JavaScript获得了一些工具来帮助解决这种问题。通过 [Web workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) 可以把一些任务交给一个名为 worker 的单独的线程，这样就可以同时运行多个JavaScript代码块。一般来说，用一个 worker 来运行一个耗时的任务，主线程就可以处理用户的交互（避免了阻塞）。

web workers 相当有用，但是他们确实也有局限。主要的一个问题是他们不能访问 DOM — 不能让一个 worker 直接更新 UI。

## 异步JavaScript

在 JavaScript 代码中，你经常会遇到两种异步编程风格：老派 callbacks，新派 promise。

### callbacks

异步 callbacks 其实就是函数，只不过是作为参数传递给那些在后台执行的其他函数。当那些后台运行的代码结束，就调用 callbacks 函数，通知你工作已经完成，或者其他有趣的事情发生了。使用 callbacks 有一点老套，在一些老派但经常使用的 API 里面，你会经常看到这种风格。

举个例子，异步callback 就是[`addEventListener()`](https://developer.mozilla.org/zh-CN/docs/Web/API/EventTarget/addEventListener)第二个参数：

```js
btn.addEventListener('click', () => {
  alert('You clicked me!');

  let pElem = document.createElement('p');
  pElem.textContent = 'This is a newly-added paragraph.';
  document.body.appendChild(pElem);
});
```

当我们把回调函数作为一个参数传递给另一个函数时，仅仅是把回调函数定义作为参数传递过去——回调函数并没有立刻执行，回调函数会在包含它的函数的某个地方异步执行，包含函数负责在合适的时候执行回调函数。



你可以自己写一个容易的，包含回调函数的函数。来看另外一个例子，用 XMLHttpRequest API （运行它, 源代码）加载资源：

```js
function loadAsset(url, type, callback) {
  let xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.responseType = type;

  xhr.onload = function() {
    callback(xhr.response);
  };

  xhr.send();
}

function displayImage(blob) {
  let objectURL = URL.createObjectURL(blob);

  let image = document.createElement('img');
  image.src = objectURL;
  document.body.appendChild(image);
}

loadAsset('coffee.jpg', 'blob', displayImage);
```

请注意，不是所有的回调函数都是异步的 — 有一些是同步的。一个例子就是使用 `Array.prototype.forEach()` 来遍历数组：

```js
const gods = ['Apollo', 'Artemis', 'Ares', 'Zeus'];

gods.forEach(function (eachName, index){
  console.log(index + '. ' + eachName);
});
```



### promise

promise 是新派的异步代码，现代的 web APIs 经常用到。 `fetch()` API就是一个很好的例子, 它基本上就是一个现代版的，更高效的 [`XMLHttpRequest`](https://developer.mozilla.org/zh-CN/docs/Web/API/XMLHttpRequest)。看个例子，来自于文章 [Fetching data from the server](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Client-side_web_APIs/Fetching_data) ：

```js
fetch('products.json').then(function(response) {
  return response.json();
}).then(function(json) {
  products = json;
  initialize();
}).catch(function(err) {
  console.log('Fetch problem: ' + err.message);
});
```

这里 `fetch()` 只需要一个参数——资源的网络 URL，返回一个 promise。**promise 是表示异步操作完成或失败的对象，可以说，它代表了一种中间状态**。 本质上，这是浏览器说“我保证尽快给您答复”的方式，因此得名“promise”。

这个概念需要练习来适应，它感觉有点像运行中的薛定谔猫。这两种可能的结果都还没有发生，因此 fetch 操作目前正在等待浏览器试图在将来某个时候完成该操作的结果。然后我们有三个代码块链接到 `fetch()` 的末尾：

- 两个 `then()` 块。两者都包含一个回调函数，如果前一个操作成功，该函数将运行，并且每个回调都接收前一个成功操作的结果作为输入，因此您可以继续对它执行其他操作。每个 `.then()` 块返回另一个 promise，这意味着可以将多个 `.then()` 块链接到另一个块上，这样就可以依次执行多个异步操作。
- 如果其中任何一个 `then()` 块失败，则在末尾运行 `catch()` 块（与同步`try...catch`类似，`catch()`提供了一个错误对象，可用来报告发生的错误类型）。



像 promise 这样的异步操作被放入**事件队列**中，事件队列在主线程完成处理后运行，这样它们就不会阻止后续 JavaScript 代码的运行。排队操作将尽快完成，然后将结果返回到 JavaScript 环境。



promise 对比 callbacks

promise与旧式callbacks有一些相似之处。它们本质上是一个返回的对象，您可以将回调函数附加到该对象上，而不必将回调作为参数传递给另一个函数。

然而，`promise` 是专门为异步操作而设计的，与旧式回调相比具有许多优点：

- 您可以使用多个 `then()` 操作将多个异步操作链接在一起，并将其中一个操作的结果作为输入传递给下一个操作。这种链接方式对回调来说要难得多，会使回调以混乱的“末日金字塔”告终 （也称为[回调地狱](http://callbackhell.com/)）。
- `promise`总是严格按照它们放置在事件队列中的顺序调用。
- 错误处理要好得多——所有的错误都由块末尾的一个 `.catch()` 块处理，而不是在“金字塔”的每一层单独处理。



## 超时和间隔

很长一段时间以来，web 平台为 JavaScript 程序员提供了许多函数，这些函数允许您在一段时间间隔过后异步执行代码，或者重复异步执行代码块，直到您告诉它停止为止，这些函数是：

- `setTimeout()`
  在指定的时间后执行一段代码。
- `setInterval()`
  以固定的时间间隔，重复运行一段代码。
- `requestAnimationFrame()`
  `setInterval()` 的现代版本，在浏览器下一次重新绘制显示之前执行指定的代码块，从而允许动画在适当的帧率下运行，而不管它在什么环境中运行。

### `setTimeout()`

正如前述， `setTimeout()` 在指定的时间后执行一段特定代码. 它需要如下参数:

- 要运行的函数，或者函数引用。

- 表示在执行代码之前等待的时间间隔的数字（以毫秒为单位），使用 0 用作 `setTimeout()` 的回调函数会立刻执行，但是在主线程代码运行之后执行。

  如果您希望设置一个代码块以便在所有主线程完成运行后立即运行，这将很有用。将其放在异步事件循环中，这样它将随后直接运行。

- 更多的参数：在指定函数运行时，希望传递给函数的值。

```js
let myGreeting = setTimeout(function() {
  alert('Hello, Mr. Universe!');
}, 2000)
```

我们指定的函数不必是匿名的。我们可以给函数一个名称，甚至可以在其他地方定义它，并将函数引用传递给 `setTimeout()` 。以下两个版本的代码片段相当于第一个版本：

```js
// With a named function
let myGreeting = setTimeout(function sayHi() {
  alert('Hello, Mr. Universe!');
}, 2000)

// With a function defined separately
function sayHi() {
  alert('Hello Mr. Universe!');
}

let myGreeting = setTimeout(sayHi, 2000);
```

传递参数给 `setTimeout()`

```js
function sayHi(who) {
  alert('Hello ' + who + '!');
}

let myGreeting = setTimeout(sayHi, 2000, 'Mr. Universe');
```



清除超时

最后，如果创建了 timeout，您可以通过调用 `clearTimeout()`，将 `setTimeout()` 调用的标识符作为参数传递给它，从而在超时运行之前取消。要取消上面的超时，你需要这样做：

```js
clearTimeout(myGreeting);
```

### `setInterval()`

当我们需要在一段时间之后运行一次代码时，`setTimeout()` 可以很好地工作。但是当我们需要反复运行代码时会发生什么，例如在动画的情况下?

让我们看一个例子。下面的函数创建一个新的Date()对象，使用 `toLocaleTimeString()` 从中提取一个时间字符串，然后在 UI 中显示它。然后，我们使用`setInterval()`  每秒运行该函数一次，创建一个每秒更新一次的数字时钟的效果。

```js
function displayTime() {
   let date = new Date();
   let time = date.toLocaleTimeString();
   document.getElementById('demo').textContent = time;
}

const createClock = setInterval(displayTime, 1000);
```

清除 intervals

`setInterval()` 永远保持运行任务,除非我们做点什么——我们可能会想阻止这样的任务，则当浏览器无法完成任何进一步的任务时我们可能得到错误，或者动画处理已经完成了。我们可以用与停止超时相同的方法来实现这一点——通过将 `setInterval()` 调用返回的标识符传递给 `clearInterval()`函数：

```js
const myInterval = setInterval(myFunction, 2000);

clearInterval(myInterval);
```

### ` requestAnimationFrame()`

`requestAnimationFrame()` 是一个专门的循环函数，旨在浏览器中高效运行动画。它基本上是现代版本的 `setInterval()` —— **它在浏览器重新加载显示内容之前执行指定的代码块**，从而允许动画以适当的帧速率运行，不管其运行的环境如何。

它是针对 `setInterval()` 遇到的问题创建的，比如 `setInterval()` 并不是针对设备优化的帧率运行，有时会丢帧。还有即使该选项卡不是活动的选项卡或动画滚出页面等问题 。

该方法将重新加载页面之前要调用的回调函数作为参数。这是您将看到的常见表达：

```js
function draw() {
   // Drawing code goes here
   requestAnimationFrame(draw);
}

draw();
```

## promise

本质上，promise 是一个对象，代表**操作的中间状态** —— 正如它的单词含义 '承诺' ，它保证在未来可能返回某种结果。虽然 Promise 并不保证操作在何时完成并返回结果，但是它保证当结果可用时，你的代码能正确处理结果，当结果不可用时，你的代码同样会被执行，来优雅的处理错误。



### 回调函数的麻烦

我们来谈谈订购披萨作为类比。为了使你的订单成功，你必须按顺序执行，不按顺序执行或上一步没完成就执行下一步是不会成功的：

1. 选择配料。如果你是优柔寡断，这可能需要一段时间，如果你无法下定决心或者决定换咖喱，可能会失败。
2. 下订单。返回比萨饼可能需要一段时间，如果餐厅没有烹饪所需的配料，可能会失败。
3. 然后你收集你的披萨吃。如果你忘记了自己的钱包，那么这可能会失败，所以无法支付比萨饼的费用！

对于旧式 callbacks，上述功能的伪代码表示可能如下所示：

```js
chooseToppings(function(toppings) {
  placeOrder(toppings, function(order) {
    collectOrder(order, function(pizza) {
      eatPizza(pizza);
    }, failureCallback);
  }, failureCallback);
}, failureCallback);
```

这很麻烦且难以阅读（通常称为“回调地狱”），需要多次调用 `failureCallback()`（每个嵌套函数一次），还有其他问题。

使用 promise改良：

```js
chooseToppings()
.then(function(toppings) {
  return placeOrder(toppings);
})
.then(function(order) {
  return collectOrder(order);
})
.then(function(pizza) {
  eatPizza(pizza);
})
.catch(failureCallback);
```

使用箭头函数，你可以进一步简化代码：

```js
chooseToppings()
.then(toppings =>
  placeOrder(toppings)
)
.then(order =>
  collectOrder(order)
)
.then(pizza =>
  eatPizza(pizza)
)
.catch(failureCallback);
```

甚至这样：

```js
chooseToppings().then(placeOrder).then(collectOrder).then(eatPizza).catch(failureCallback);
```

你甚至可以这样做，因为函数只是直接传递它们的参数，所以不需要额外的函数层，但是，这并不容易阅读，如果你的块比我们在此处显示的更复杂，则此语法可能无法使用。



promise 很重要，因为大多数现代 Web API 都将它们用于执行潜在冗长任务的函数。要使用现代 Web 技术，你需要使用 promise。

在第一个示例中，我们将使用`fetch()`方法从Web获取图像，`blob()` 方法来转换获取响应的原始内容到 [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob) 对象，然后在 [ ](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img)元素内显示该`blob`。

```js
let promise = fetch('coffee.jpg');
```

为了响应成功完成操作（在这种情况下，当返回 Response 时），我们调用 promise 对象的 `.then()` 方法。 `.then()` 块中的回调（称为执行程序）仅在promise 调用成功完成时运行并返回 `Response` 对象，它将返回的 Response 对象作为参数传递。

我们立即对此响应运行`blob()`方法以确保响应主体完全下载，并且当它可用时将其转换为我们可以执行某些操作的`Blob`对象。返回的结果如下：

```js
response => response.blob()
```

这是下面的简写

```js
function(response) {
    return response.blob();
}
```

好的，我们还需要做点额外的工作。Fetch promise不会产生 404 或 500错误，只有在产生像网路故障的情况时才会不工作。总的来说，Fetch promise总是成功运行，即使 [response.ok](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok)  属性是` false`。为了产生404错误，我们需要判断 `response.ok` ，如果是 `false`，抛出错误，否则返回 blob。就像下面的代码这样做。

```js
let promise2 = promise.then(response => {
  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  } else {
    return response.blob();
  }
});
```

每次调用 `.then()` 都会创建一个新的 promise。这非常有用，因为 `blob()` 方法也返回一个 promise，我们可以通过调用第二个 promise 的 `.then()` 方法来处理它在履行时返回的 Blob 对象。因为我们想要对 blob 执行一些更复杂的操作，而不仅仅运行单个方法并返回结果，这次我们需要将函数体包装成花括号。

```js
let promise3 = promise2.then(myBlob => {
    let objectURL = URL.createObjectURL(myBlob);
    let image = document.createElement('img');
    image.src = objectURL;
    document.body.appendChild(image);
})
```

最后，我们可以通过运行前一个 promise 的 `.catch()` 方法来添加错误处理：

```js
let errorCase = promise3.catch(e => {
  console.log('There has been a problem with your fetch operation: ' + e.message);
});
```



将代码块链在一起，如下：

```js
fetch('coffee.jpg')
.then(response => {
  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  } else {
    return response.blob();
  }
})
.then(myBlob => {
  let objectURL = URL.createObjectURL(myBlob);
  let image = document.createElement('img');
  image.src = objectURL;
  document.body.appendChild(image);
})
.catch(e => {
  console.log('There has been a problem with your fetch operation: ' + e.message);
});
```



### promise 术语

1. 创建 promise 时，它既不是成功也不是失败状态。这个状态叫作 **pending**（待定）。如：

   ```js
   let promise = fetch('coffee.jpg');
   ```

2. 当 promise 返回时，称为 **resolved**（已解决）；

   1. 一个成功 **resolved** 的 promise 称为 **fullfilled**（**实现**）。它返回一个值，可以通过将 `.then()` 块链接到 promise 链的末尾来访问该值。` .then() `块中的执行程序函数将访问 promise 的返回值。
   2. 一个不成功 **resolved** 的 promise 被称为 **rejected**（**拒绝**）。它返回一个原因（**reason**），一条错误消息，说明为什么拒绝 promise。可以通过将 `.catch()` 块链接到 promise 链的末尾来访问此原因。



链接进程一个接一个地发生都很好，但是如果你想在一大堆 promise 全部完成之后运行一些代码呢？

你可以使用 `Promise.all()` 静态方法完成此操作。这将一个 promise 数组作为输入参数，并返回一个新的 promise 对象，只有当数组中的所有 promise 都满足时才会满足。它看起来像这样：

```js
Promise.all([a, b, c]).then(values => {
  ...
});
```

如果它们都**实现**，那么数组中的结果将作为参数传递给`.then()`块中的执行器函数。如果传递给`Promise.all()`的任何一个 promise **拒绝**，整个块将**拒绝**。



### 运行最终代码

在 promise 完成后，你可能希望运行最后一段代码，无论它是否已实现（fullfilled）或被拒绝（rejected）。此前，你必须在 `.then()` 和 `.catch()` 回调中包含相同的代码，例如：

```js
myPromise
.then(response => {
  doSomething(response);
  runFinalCode();
})
.catch(e => {
  returnError(e);
  runFinalCode();
});
```

在现代浏览器中，`.finally()` 方法可用，它可以链接到常规 promise 链的末尾，允许你减少代码重复并更优雅地执行操作。上面的代码现在可以写成如下：

```js
myPromise
.then(response => {
  doSomething(response);
})
.catch(e => {
  returnError(e);
})
.finally(() => {
  runFinalCode();
});
```



### 构建自定义 promise

可以使用 `Promise()` 构造函数构建自己的 promise。当你需要使用现有的旧项目代码、库或框架以及基于现代 promise 的代码时，这会派上用场。比如，当你遇到没有使用 promise 的旧式异步 API 的代码时，你可以用 promise 来重构这段异步代码。

让我们看一个简单的示例来帮助你入门 —— 这里我们用 promise 包装一了个`setTimeout()，`它会在两秒后运行一个函数，该函数将用字符串“Success!”。

```js
let timeoutPromise = new Promise((resolve, reject) => {
  setTimeout(function(){
    resolve('Success!');
  }, 2000);
});
```

`resolve() `和 `reject()`是用来**实现**和**拒绝**新创建的 promise 的函数。此处，promise 成功运行通过显示字符串“Success!”。

因此，当你调用此 promise 时，可以将.then()块链接到它的末尾，它将传递给 `.then()` 块一串“Success!”。在下面的代码中，我们显示出该消息：

```js
timeoutPromise
.then((message) => {
   alert(message);
})
```

> **注意**: 这里使用`resolve()` 方法实现 promise  ，而不是 `fullfill()`。

拒绝一个自定义 promise，我们可以创建一个 `reject()` 方法拒绝 promise  —— 就像 `resolve()` 一样，这需要一个值，但在这种情况下，它是拒绝的原因，即将传递给 `.catch()` 的错误块。

```js
function timeoutPromise(message, interval) {
  return new Promise((resolve, reject) => {
    if (message === '' || typeof message !== 'string') {
      reject('Message is empty or not a string');
    } else if (interval < 0 || typeof interval !== 'number') {
      reject('Interval is negative or not a number');
    } else {
      setTimeout(function(){
        resolve(message);
      }, interval);
    }
  });
};
```

由于`timeoutPromise()`函数返回一个`Promise`，我们可以将`.then()`，`.catch()`等链接到它上面以利用它的功能。

```js
timeoutPromise('Hello there!', 1000)
.then(message => {
   alert(message);
})
.catch(e => {
  console.log('Error: ' + e);
});
```



## async 和 await

### async

`async` 关键字，把它放在函数声明之前，使其成为 [async function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function)。异步函数是一个知道怎样使用 `await` 关键字调用异步代码的函数。

```js
function hello() { return "Hello" };
hello();
```

```js
async function hello() { return "Hello" };
hello();
```

异步函数会返回一个 promise。这是异步函数的特征之一 —— 它保证函数的返回值为 promise。

你可以使用箭头函数：

```js
let hello = async () => { return "Hello" };
```

要实际使用 promise 完成时返回的值，我们可以使用 `.then()` 块：

```js
hello().then((value) => console.log(value))
```



### await

当 [await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await) 关键字与异步函数一起使用时，它的真正优势就变得明显了 —— 事实上， **await 只在异步函数里面才起作用**。它可以放在任何异步的，基于 promise 的函数之前。它会暂停代码在该行上，直到 promise 完成，然后返回结果值。在暂停的同时，其他正在等待执行的代码就有机会执行了。



让我们回顾一下我们在上一篇文章中简单的 fetch 示例：

```js
fetch('coffee.jpg')
.then(response => response.blob())
.then(myBlob => {
  let objectURL = URL.createObjectURL(myBlob);
  let image = document.createElement('img');
  image.src = objectURL;
  document.body.appendChild(image);
})
.catch(e => {
  console.log('There has been a problem with your fetch operation: ' + e.message);
});
```

使用 async 和 await 关键字后我们可以这样改造上述代码：

```js
async function myFetch() {
  // 线程会暂停在 await 修饰的代码行上，直到该行的 promise 完成
  let response = await fetch('coffee.jpg');
  let myBlob = await response.blob();

  let objectURL = URL.createObjectURL(myBlob);
  let image = document.createElement('img');
  image.src = objectURL;
  document.body.appendChild(image);
}

myFetch()
.catch(e => {
  console.log('There has been a problem with your fetch operation: ' + e.message);
});
```

由于 `async` 关键字将函数转换为 promise，您可以重构以上代码 —— 使用 promise 和 await 的混合方式，将函数的后半部分抽取到新代码块中。这样做可以更灵活：

```js
async function myFetch() {
  let response = await fetch('coffee.jpg');
  return await response.blob();
}

myFetch().then((blob) => {
  let objectURL = URL.createObjectURL(blob);
  let image = document.createElement('img');
  image.src = objectURL;
  document.body.appendChild(image);
})
.catch((e) =>
  console.log('There has been a problem with your fetch operation: ' + e.message);
);
```



使用了 await 之后，您可以将同步的 try...catch 结构和 async/await 一起使用 ：

```js
async function myFetch() {
  try {
    let response = await fetch('coffee.jpg');
    let myBlob = await response.blob();

    let objectURL = URL.createObjectURL(myBlob);
    let image = document.createElement('img');
    image.src = objectURL;
    document.body.appendChild(image);
  } catch(e) {
    console.log(e);
  }
}

myFetch();
```

如果你想使用我们上面展示的第二个（重构）代码版本，你最好继续混合方式并将 `.catch()` 块链接到 `.then()` 调用的末尾，就像这样：



### async/await 缺陷

`async/await` 让你的代码看起来是同步的，在某种程度上，也使得它的行为更加地同步。 `await` 关键字会阻塞其后的代码，直到 promise 完成，就像执行同步操作一样。它确实可以允许其他任务在此期间继续运行，但您自己的代码被阻塞。

这意味着您的代码可能会因为大量 `await` 的 promise 相继发生而变慢。每个`await` 都会等待前一个完成，而你实际想要的是所有的这些 promise 同时开始处（就像我们没有使用`async/await`时那样）。

有一种模式可以缓解这个问题——通过将 promise 对象存储在变量中来同时开始它们，然后等待它们全部执行完毕。



```js
function timeoutPromise(interval) {
  return new Promise((resolve, reject) => {
    setTimeout(function(){
      resolve("done");
    }, interval);
  });
};
```



### async/await 类方法

最后值得一提的是，我们可以在类/对象方法前面添加 async，以使它们返回 promise，并 await 它们内部的 promise。

```js
class Person {
  constructor(first, last, age, gender, interests) {
    this.name = {
      first,
      last
    };
    this.age = age;
    this.gender = gender;
    this.interests = interests;
  }

  async greeting() {
    return await Promise.resolve(`Hi! I'm ${this.name.first}`);
  };

  farewell() {
    console.log(`${this.name.first} has left the building. Bye for now!`);
  };
}

let han = new Person('Han', 'Solo', 25, 'male', ['Smuggling']);
```

